### DISCUSSION TITLE
Calculate the number of lone pairs per atom

### DISCUSSION THREAD

**Discussion by ghiander** [rdkit/rdkit#4027] [Q&A]
Hi all,

I was wondering whether you have got already an efficient way to calculate the number of lone pairs of an atom object.
I am currently playing with this test molecule: CSCc1ccc(CNC(=O)C)c(CCC(F)(Cl)Br)n1

Giammy

**Comment by dehaenw** [comment]
Within first approximation, you can say that the amount of unpaired electrons is:

$n_e=\lfloor\frac{v_e-c-\sum(b)}{2}\rfloor$

with $n_e$ the amount of lone pairs, $\lfloor x \rfloor$ rounding $x$ down to the nearest integer $v_e$ the valence electrons, $c$ the formal charge and $\sum(b)$ the sum of bond orders. Of course, this does not account for radicals, carbenes etc, but for a typical druglike molecule probably suffices.

This is easily implemented in rdkit:
```python
from rdkit import Chem
import math

tbl = Chem.GetPeriodicTable()

def count_lone_pairs(a):
    v=tbl.GetNOuterElecs(a.GetAtomicNum())
    c=a.GetFormalCharge()
    b=sum([bond.GetBondTypeAsDouble() for bond in a.GetBonds()])
    return math.floor(0.5*(v-c-b))
```

And counting the total amount of lone pairs in a given molecule is then (note how explicit Hs need to be added for the formula to work correctly):

```python
def sum_lone_pairs(mol):
    mol_h=Chem.AddHs(mol)
    return sum([count_lone_pairs(a) for a in mol_h.GetAtoms()])
    
    
m=Chem.MolFromSmiles("CSCc1ccc(CNC(=O)C)c(CCC(F)(Cl)Br)n1")
sum_lone_pairs(m)
```

which gives a count of 15 for your molecule.

**Reply by dehaenw** [reply]
Ha, I didn't see this post is 2 years old. Hopefully it may be useful anyway.

**Comment by UnixJunkie** [comment]
looks like a useful descriptors; if I am not mistaken, this is required by some QM code.
Could be added at the atom and molecule level in rdkit.
